src/java_bytecode/generate_java_generic_method.cpp

```diff
@@ -0,0 +1,296 @@
+
+  // Handle a generic return type
+  typet &method_return_type=specialized_code.return_type();
+  if(is_java_generic_parameter(method_return_type))
+  {
+    instantiate_java_generic_parameter(
+      to_java_generic_parameter(method_return_type),
+      concrete_type_map);
+  }
+  else if(is_java_generic_type(method_return_type))
+  {
+    instantiate_java_generic_type(
+      to_java_generic_type(method_return_type),
+      concrete_type_map);
+  }
+
+  // Set the name of the specialized method, but ensure we remember the
+  // original method name because we need it for cleaning up internal symbol
+  // names...
+  const irep_idt original_name=specialized_method.name;
+  specialized_method.name=
+    id2string(specialized_method.name)+signature_decoration;
+
+  for(code_typet::parametert &parameter : specialized_code.parameters())
+  {
+    // Update the symbol name of the parameter to match the newly decorated
+    // name of the specialized method.
+    decorate_identifier(parameter, ID_C_identifier, original_name,
+      signature_decoration);
+
+    typet &parameter_type=parameter.type();
+
+    if(is_java_generic_parameter(parameter_type))
+    {
+      instantiate_java_generic_parameter(
+        to_java_generic_parameter(parameter_type),
+        concrete_type_map);
+    }
+    else if(is_java_generic_type(parameter_type))
+    {
+      instantiate_java_generic_type(
+        to_java_generic_type(parameter_type),
+        concrete_type_map);
+    }
+  }
+
+  // Updated the body of the specialized method so that all references to
+  // symbols in the method body refer to the specialized method and not the
+  // generic method.
+  // FIXME: Probably also need to update references to fields to point to the
+  // FIXME: specialized class fields, rather than the generic class fields?
+  auto body_iterator=specialized_method.value.depth_begin();
+  const auto body_end=specialized_method.value.depth_end();
+
+  while(body_iterator!=body_end)
+  {
+    if(body_iterator->id()==ID_symbol)
+    {
+      exprt &symbol_expr=body_iterator.mutate();
+      typet &symbol_type=symbol_expr.type();
+      if(is_java_generic_parameter(symbol_type))
+      {
+        instantiate_java_generic_parameter(
+          to_java_generic_parameter(symbol_type),
+          concrete_type_map);
+      }
+      else if(is_java_generic_type(symbol_type))
+      {
+        instantiate_java_generic_type(
+          to_java_generic_type(symbol_type),
+          concrete_type_map);
+      }
+      decorate_identifier(symbol_expr, ID_identifier, original_name,
+        signature_decoration);
+    }
+
+    ++body_iterator;
+  }
+
+  if(symbol_table.add(specialized_method))
+  {
+    // We should never hit this because the first thing
+    // we do before starting to specialize the method
+    // is check that the proposed symbol name doesn't
+    // already exist.
+    throw "specialized method symbol already exists";
```

Comment:
Probably better to use a structured exception as I feel like we can probably catch problems in generics and essentially try without the specialisation